[Java FFI][Bug 18846] Implement a parser for the Java FFI DSL

livecodeali opened the pull request:
Replaces https://github.com/livecode/livecode/pull/4438

livecodeali committed:
[[ Java FFI ]] Add Java FFI DSL parser
This parser is written in gentle and is largely based on
lc-compile. The main difference is the fact that names are
qualified, and accordingly '.' is not allowed in name literals.

livecodeali committed:
[[ Java FFI ]] Add Java FFI DSL check output
Add a --check option to the java-dsl-parse program which combines
the input files into one and outputs to the specified file. This
is used to check the idempotence of the parsing step.

livecodeali committed:
[[ Java FFI ]] Add parser checking and tests
Add a java dsl parser test runner (base on the lc-compile one) and
add tests.

livecodeali committed:
[[ Java FFI ]] Add LCB output
Add an --output option to the java-dsl-parse program which outputs
LCB glue code to the specified location. The glue code enables the
methods, constructors, variables and constants of the java class
described in the target DSL files to be accessed from LCB.

livecodeali committed:
[[ Java FFI DSL ]] Generate constructor code properly
Call the to-be-implemented java.lcb module code when generating
constructor glue.

livecodeali committed:
[[ Java FFI DSL ]] Add placeholders to deal with variadic args
We don't yet support variadic args in LCB, so use a list for the
time being. Calling such functions through the JNI will not work
until the relevant support is added to the java.lcb libscript module.

livecodeali committed:
[[ Java FFI DSL ]] Ensure params are applied correctly
Fix an issue whereby parameters qualified names could not be
retrieved.

livecodeali committed:
[[ Java FFI DSL ]] Check template types parameters match
Add a check to ensure generic type parameters match properly.

livecodeali committed:
[[ Java FFI DSL ]] Allow unbounded wildcards in named type parameters

livecodeali committed:
[[ Java FFI DSL ]] Add docs for java dsl parser

peter-b commented on toolchain/java-dsl-parse/src/LAYOUT.b:
```diff
@@ -0,0 +1,2 @@
+\  { AdvanceCurrentPosition(1); }
+\t {  }
```
You need to `AdvanceCurrentPosition(1);` for tab characters, or column numbers will be reported incorrectly when errors occur on lines indented with tabs.

peter-b commented on toolchain/java-dsl-parse/src/DOUBLE_LITERAL.t:
```diff
@@ -0,0 +1,6 @@
+[0-9]+"."[0-9]+([eE][-+]?[0-9]+)? {
```
It looks like `1e1` is not parsed as a numeric literal, although it is a numeric literal in LCB.  This isn't necessarily a problem, but it's probably something to be aware of.

livecodeian committed:
[[ Java FFI ]] Refactor common code between lc-compile / java-ffi-parse into libcompile

livecodeali committed:
Merge pull request #6 from livecodeian/feature-java_ffi_dsl_parser_libcompile

[[ Java FFI ]] Refactor common code between lc-compile / java-ffi-parse into libcompile

livecodeali committed:
Merge branch 'develop' of github.com:livecode/livecode into feature-java_ffi_dsl_parser

livecodeali committed:
[[ Java DSL Parser ]] Use Define/HasUnqualifiedMeaning functions
At the moment, the DSL parser just uses qualified names everywhere
behind the scenes, so it should use DefineUnqualifiedMeaning and
HasUnqualifiedMeaning functions (HasMeaning and DefineMeaning take
a namespace as a parameter)

livecodeali committed:
[[ Java DSL Parser ]] LCB output module name to be specified on the command line
Since lcb-java files are likely to be re-used, and the output is a single lcb
file, it makes sense for the lcb module name to be specified on the command line

livecodeali committed:
[[ Java FFI DSL ]] Add test for circular references across package definitions

livecodeali committed:
[[ Tests ]] Refactor compiler test runner
This patch continues the attempt to structure test runners
'polymorphically' using behaviors. A compiler test runner
inherits from the main test runner, but each compiler test
runner requires its own implementation of
- whether the output from the 'run' command satisfies the assertion
- command line syntax error message
- running a test
- returning the relevant file extension

livecodeali committed:
[[ Java FFI DSL ]] Use refactored test runner with java dsl parser tests

livecodeali committed:
[[ Tests ]] Specify log file as environment variable

livecodeali committed:
[[ LC Compile FFI Java ]] Rename java ffi dsl tool

livecodeali committed:
[[ LC Compile FFI Java ]] Fix memory leak

runrevmark commented:
@livecode-vulcan review ok 4468582

livecode-vulcan commented:
💙 review by @runrevmark ok 4468582e5f1b27b87644b8ae4839c35b087f5997

livecode-vulcan commented:
😞 test failure 4468582e5f1b27b87644b8ae4839c35b087f5997

* try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/998)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1004)
* try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/996)
* try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/997)
* try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1004)
* try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1018)
* try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/995)

livecodeali committed:
[[ LC Compile FFI Java ]] Declare vars at beginning of handler

livecodeali committed:
[[ LC Compile Java FFI ]] Add constructor binding glue code
We need to be able to use the general type bridging system to call
constructors with parameters easily, so this patch adds glue code
for constructors which adds a foreign binding string of the form
"java:qualified.class.name>new(paramString)"

Correspondingly, glue code is generated to call constructors with
the appropriate types and return the constructed java object.

livecodeali committed:
[[ LC Compile FFI Java ]] Add glue code for byteArray to Data conversion
This patch adds a special case to generate glue code for converting
java byte arrays to Data and back. This means that it is not
necessary to hard-code the conversion in libfoundation, so that in
future better glue code could be generated to elide conversion when
java byteArrays are passed around.

livecodeali committed:
[[ LC Compile FFI Java ]] Ensure generated glue code is correct
- Ensure static methods and fields are marked as such in the
binding string using !static
- Generate getters for variables and constants, and setters for
non-final variables

peter-b commented on toolchain/lc-compile-ffi-java/src/LAYOUT.b:
```diff
@@ -0,0 +1,2 @@
+\  { AdvanceCurrentPosition(1); }
+\t {  }
```
You need to `AdvanceCurrentPosition` for tab characters, or the incorrect column number will be reported in error and warning messages.

livecodeali replied:
Good catch :-)

peter-b reviewed (changes requested):
Oops, I found one minor issue that I didn't flag up properly earlier.

livecodeali committed:
[[ LC Compile FFI Java ]] Advance position for tab character

peter-b commented:
@livecode-vulcan review ok 5e452a7

livecode-vulcan commented:
💙 review by @peter-b ok 5e452a723a2d9b7856c99b09bfb4bae7b7a27035

livecode-vulcan commented:
😞 test failure 5e452a723a2d9b7856c99b09bfb4bae7b7a27035

* try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1003)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1009)
* try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1001)
* try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1002)
* try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1009)
* try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1023)
* try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1000)

livecodeali committed:
[[ LC Cpmpile FFI Java ]] Always define yywrap

peter-b commented:
@livecode-vulcan review ok 33f63c2

livecode-vulcan commented:
💙 review by @peter-b ok 33f63c2eecfb71fde4d30e59d9803a90948ec522

livecode-vulcan commented:
😎 test success 33f63c2eecfb71fde4d30e59d9803a90948ec522

* try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1004)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1010)
* try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1002)
* try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1003)
* try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1010)
* try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1024)
* try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1001)
